File: shared/utils/logger.ts
/**
 * Logging utilities for Lambda functions
 */

export enum LogLevel {
  DEBUG = 'DEBUG',
  INFO = 'INFO',
  WARN = 'WARN',
  ERROR = 'ERROR'
}

export interface LogContext {
  requestId?: string;
  functionName?: string;
  userId?: string;
  itemId?: string;
  [key: string]: any;
}

export class Logger {
  private context: LogContext;
  private logLevel: LogLevel;

  constructor(context: LogContext = {}, logLevel: LogLevel = LogLevel.INFO) {
    this.context = context;
    this.logLevel = logLevel;
  }

  private shouldLog(level: LogLevel): boolean {
    const levels = [LogLevel.DEBUG, LogLevel.INFO, LogLevel.WARN, LogLevel.ERROR];
    return levels.indexOf(level) >= levels.indexOf(this.logLevel);
  }

  private formatMessage(level: LogLevel, message: string, meta?: any): string {
    const timestamp = new Date().toISOString();
    const logEntry = {
      timestamp,
      level,
      message,
      context: this.context,
      ...(meta && { meta })
    };
    return JSON.stringify(logEntry);
  }

  debug(message: string, meta?: any): void {
    if (this.shouldLog(LogLevel.DEBUG)) {
      console.log(this.formatMessage(LogLevel.DEBUG, message, meta));
    }
  }

  info(message: string, meta?: any): void {
    if (this.shouldLog(LogLevel.INFO)) {
      console.log(this.formatMessage(LogLevel.INFO, message, meta));
    }
  }

  warn(message: string, meta?: any): void {
    if (this.shouldLog(LogLevel.WARN)) {
      console.warn(this.formatMessage(LogLevel.WARN, message, meta));
    }
  }

  error(message: string, error?: Error, meta?: any): void {
    if (this.shouldLog(LogLevel.ERROR)) {
      const errorMeta = {
        ...meta,
        ...(error && {
          error: {
            name: error.name,
            message: error.message,
            stack: error.stack
          }
        })
      };
      console.error(this.formatMessage(LogLevel.ERROR, message, errorMeta));
    }
  }

  withContext(additionalContext: LogContext): Logger {
    return new Logger({ ...this.context, ...additionalContext }, this.logLevel);
  }

  setLogLevel(level: LogLevel): void {
    this.logLevel = level;
  }
}

// Default logger instance
export const logger = new Logger({
  functionName: process.env.AWS_LAMBDA_FUNCTION_NAME,
}, process.env.LOG_LEVEL as LogLevel || LogLevel.INFO);

// Helper function to create a logger with request context
export function createRequestLogger(requestId: string, additionalContext: LogContext = {}): Logger {
  return logger.withContext({ requestId, ...additionalContext });
}

File: functions/delete-item/logger.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"time"
)

// LogLevel represents the severity level of a log entry
type LogLevel string

const (
	DEBUG LogLevel = "DEBUG"
	INFO  LogLevel = "INFO"
	WARN  LogLevel = "WARN"
	ERROR LogLevel = "ERROR"
)

// LogContext represents the context information for logging
type LogContext struct {
	RequestID    string `json:"requestId,omitempty"`
	FunctionName string `json:"functionName,omitempty"`
	Operation    string `json:"operation,omitempty"`
}

// Logger provides structured logging functionality
type Logger struct {
	context  LogContext
	logLevel LogLevel
}

// LogEntry represents a structured log entry
type LogEntry struct {
	Timestamp string                 `json:"timestamp"`
	Level     LogLevel               `json:"level"`
	Message   string                 `json:"message"`
	Context   LogContext             `json:"context"`
	Meta      map[string]interface{} `json:"meta,omitempty"`
	Error     *ErrorInfo             `json:"error,omitempty"`
}

// ErrorInfo represents error information in logs
type ErrorInfo struct {
	Name    string `json:"name"`
	Message string `json:"message"`
	Stack   string `json:"stack,omitempty"`
}

// NewLogger creates a new logger with the given context
func NewLogger(context LogContext, logLevel LogLevel) *Logger {
	if logLevel == "" {
		logLevel = INFO
	}
	return &Logger{
		context:  context,
		logLevel: logLevel,
	}
}

// shouldLog determines if a message should be logged based on level
func (l *Logger) shouldLog(level LogLevel) bool {
	levels := map[LogLevel]int{
		DEBUG: 0,
		INFO:  1,
		WARN:  2,
		ERROR: 3,
	}
	return levels[level] >= levels[l.logLevel]
}

// log writes a structured log entry
func (l *Logger) log(level LogLevel, message string, meta map[string]interface{}, err error) {
	if !l.shouldLog(level) {
		return
	}

	entry := LogEntry{
		Timestamp: time.Now().UTC().Format(time.RFC3339),
		Level:     level,
		Message:   message,
		Context:   l.context,
		Meta:      meta,
	}

	if err != nil {
		entry.Error = &ErrorInfo{
			Name:    "Error",
			Message: err.Error(),
		}
	}

	jsonLog, _ := json.Marshal(entry)
	fmt.Println(string(jsonLog))
}

// Debug logs a debug message
func (l *Logger) Debug(message string, meta map[string]interface{}) {
	l.log(DEBUG, message, meta, nil)
}

// Info logs an info message
func (l *Logger) Info(message string, meta map[string]interface{}) {
	l.log(INFO, message, meta, nil)
}

// Warn logs a warning message
func (l *Logger) Warn(message string, meta map[string]interface{}) {
	l.log(WARN, message, meta, nil)
}

// Error logs an error message
func (l *Logger) Error(message string, err error, meta map[string]interface{}) {
	l.log(ERROR, message, meta, err)
}

// WithContext creates a new logger with additional context
func (l *Logger) WithContext(additionalContext LogContext) *Logger {
	newContext := l.context
	if additionalContext.RequestID != "" {
		newContext.RequestID = additionalContext.RequestID
	}
	if additionalContext.FunctionName != "" {
		newContext.FunctionName = additionalContext.FunctionName
	}
	if additionalContext.Operation != "" {
		newContext.Operation = additionalContext.Operation
	}

	return &Logger{
		context:  newContext,
		logLevel: l.logLevel,
	}
}

// LogDatabaseOperation logs a database operation with timing
func (l *Logger) LogDatabaseOperation(operation, table string, key map[string]interface{}, duration time.Duration) {
	meta := map[string]interface{}{
		"type":      "database",
		"operation": operation,
		"table":     table,
		"duration":  fmt.Sprintf("%dms", duration.Milliseconds()),
	}
	if key != nil {
		meta["key"] = key
	}
	l.Debug(fmt.Sprintf("DynamoDB %s", operation), meta)
}

// CreateRequestLogger creates a logger with request context
func CreateRequestLogger(requestID string, additionalContext LogContext) *Logger {
	context := LogContext{
		RequestID:    requestID,
		FunctionName: additionalContext.FunctionName,
		Operation:    additionalContext.Operation,
	}

	// Get log level from environment or default to INFO
	logLevel := INFO
	if envLevel := getEnvLogLevel(); envLevel != "" {
		logLevel = envLevel
	}

	return NewLogger(context, logLevel)
}

// getEnvLogLevel gets log level from environment variable
func getEnvLogLevel() LogLevel {
	switch os.Getenv("LOG_LEVEL") {
	case "DEBUG":
		return DEBUG
	case "INFO":
		return INFO
	case "WARN":
		return WARN
	case "ERROR":
		return ERROR
	default:
		return INFO
	}
}